week7_02.py

```diff
@@ -0,0 +1,123 @@
+            # Add bool to mark if read or not
+            'read': False
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, usernames: str, num_of_messages_to_read: int = None):
+        """
+        Return the first "num_of_messages_to_read" messages in the user's inbox.
+        If no "num_of_messages_to_read" have been forwarded, return all messages in the user's inbox.
+        The messages will be marked as read and will not be returned to the user on the next call.
+        :param usernames: Username you want to read messages
+        :param num_of_messages_to_read: (int) The maximum number of messages the user wants to read
+        :return: list_of_messages_to_return List of "num_of_messages_to_read" which have not yet been read
+        """
+
+        list_of_messages_to_return = []
+        if not num_of_messages_to_read or num_of_messages_to_read > len(self.boxes[usernames]):
+            num_of_messages_to_read = len(self.boxes[usernames])
+
+        # It could have been written shorter but it came at the expense of understanding the code
+        for message in self.boxes[usernames][:num_of_messages_to_read]:
+            if not message['read']:
+                list_of_messages_to_return.append(message)
+                message['read'] = True
+
+        return list_of_messages_to_return
+
+    def search_inbox(self, usernames: str, string: str) -> list:
+        """ Return all the messages inside user_name's box that include at their body the string.
+        :param usernames: Wanted usernames box to search in.
+        :param string: String to search for.
+        :return: List of the messages that including the string at their body.
+        :raises KeyError: If user name doesn't exist.
+        """
+
+        return [[message for message in self.boxes[usernames] if string in message["body"]]]
+
+
+if __name__ == '__main__':
+
+    users = ('aaa', 'bbb', 'ccc')
+    post_office = PostOffice(users)
+    message_id = post_office.send_message(
+        sender='aaa',
+        recipient='bbb',
+        message_body='Hello,1',
+    )
+    message_id = post_office.send_message(
+        sender='aaa',
+        recipient='bbb',
+        message_body='Hello,2',
+    )
+    message_id = post_office.send_message(
+        sender='bbb',
+        recipient='ccc',
+        message_body='Hello,3',
+    )
+    message_id = post_office.send_message(
+        sender='ccc',
+        recipient='aaa',
+        message_body='Hello,4',
+    )
+    print(f" sent message number {message_id}.")
+    print(post_office.boxes['aaa'])
+    print(post_office.boxes['bbb'])
+    print(post_office.boxes['ccc'])
+    print(post_office.read_inbox('aaa', 1))
+    print(post_office.read_inbox('bbb', 1))
+    print(post_office.read_inbox('bbb', 2))
+    print(post_office.search_inbox('bbb', "2"))
```

Comment:
File should end with a newline - 
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline